File: app/src/main/java/com/util/Gravity.java
package com.util;

import com.drawable.*;

import java.util.List;

@SuppressWarnings("unused")
public final class Gravity {
    private Drawable[] drawables;
    private double gravity;
    private int canvasWidth;
    private int canvasHeight;

    public Gravity(Drawable drawable, double gravity) {
        this(new Drawable[]{drawable}, gravity);
    }

    public Gravity(Drawable[] drawables, double gravity) {
        this.drawables = new Drawable[drawables.length];
        System.arraycopy(drawables, 0, this.drawables, 0, drawables.length);
        this.gravity = gravity;
    }

    public Gravity(List<? extends Drawable> drawables, double gravity) {
        this.drawables = new Drawable[drawables.size()];
        for (int i = 0; i < drawables.size(); i++) {
            this.drawables[i] = drawables.get(i);
        }
        this.gravity = gravity;
    }

    public Drawable[] getDrawables() {
        return drawables;
    }

    public double getGravity() {
        return gravity;
    }

    public void setGravity(double gravity) {
        this.gravity = gravity;
    }

    private void manage(Circle c) {
        if (c.getY() + c.getRadius() > canvasHeight) {
            c.setY(canvasHeight - c.getRadius());
            c.setYVelocity(c.getYVelocity() * -0.6);
        }
        if (c.getY() - c.getRadius() < 0) {
            c.setY(c.getRadius());
            c.setYVelocity(c.getYVelocity() * -0.6);
        }
        if (c.getX() + c.getRadius() > canvasWidth) {
            c.setX(canvasWidth - c.getRadius());
            c.setXVelocity(c.getXVelocity() * -0.6);
        }
        if (c.getX() - c.getRadius() < 0) {
            c.setX(c.getRadius());
            c.setXVelocity(c.getXVelocity() * -0.6);
        }
    }

    private void manage(Rectangle r) {
        if (r.getY() + r.getHeight() > canvasHeight) {
            r.setY(canvasHeight - r.getHeight());
            r.setYVelocity(r.getYVelocity() * -0.6);
        }
        if (r.getY() < 0) {
            r.setY(-r.getY());
            r.setYVelocity(r.getYVelocity() * -0.6);
        }
        if (r.getX() + r.getWidth() > canvasWidth) {
            r.setX(canvasWidth - r.getWidth());
            r.setXVelocity(r.getXVelocity() * -0.6);
        }
        if (r.getX() < 0) {
            r.setX(-r.getX());
            r.setXVelocity(r.getXVelocity() * -0.6);
        }
    }

    private void manage(Text t) {
        if (t.getY() + t.getTextSize() > canvasHeight + (t.getTextSize() * 2)) {
            t.setY(canvasHeight - (t.getTextSize() * 2));
            t.setYVelocity(t.getYVelocity() * -0.6);
        }
        if (t.getY() < 0) {
            t.setY(t.getY() + (t.getTextSize() / 2));
            t.setYVelocity(t.getYVelocity() * -0.6);
        }
        if (t.getX() + t.getTextSize() > canvasWidth + (t.getTextSize() / 2) - (t.getText().length() * 16)) {
            t.setX(canvasWidth - (t.getTextSize() / 2) - (t.getText().length() * 16));
            t.setXVelocity(t.getXVelocity() * -0.6);
        }
        if (t.getX() < 0) {
            t.setX(t.getX() + (t.getTextSize() / 2));
            t.setXVelocity(t.getXVelocity() * -0.6);
        }
    }

    private void manage(Triangle t) {
        if (t.getY() > canvasHeight) {
            t.setY(canvasHeight);
            t.setYVelocity(t.getYVelocity() * -0.6);
        }
        if (t.getY() - t.getHeight() < 0) {
            t.setY(t.getHeight());
            t.setYVelocity(t.getYVelocity() * -0.6);
        }
        if (t.getX() + t.getSideLengths()[2] > canvasWidth) {
            t.setX(canvasWidth - t.getSideLengths()[2]);
            t.setXVelocity(t.getXVelocity() * -0.6);
        }
        if (t.getX() < 0) {
            t.setX(0);
            t.setXVelocity(t.getXVelocity() * -0.6);
        }
    }

    private void manage(RightTriangle t) {
        if (t.getY() > canvasHeight) {
            t.setY(canvasHeight);
            t.setYVelocity(t.getYVelocity() * -0.6);
        }
        if (t.getY() - t.getHeight() < 0) {
            t.setY(t.getHeight());
            t.setYVelocity(t.getYVelocity() * -0.6);
        }
        if (t.getX() + t.getSideLengths()[1] > canvasWidth) {
            t.setX(canvasWidth - t.getSideLengths()[1]);
            t.setXVelocity(t.getXVelocity() * -0.6);
        }
        if (t.getX() < 0) {
            t.setX(0);
            t.setXVelocity(t.getXVelocity() * -0.6);
        }
    }

    private void manage(Line l) {
        if (l.getY() > canvasHeight) {
            l.setY(canvasHeight);
            l.setYVelocity(l.getYVelocity() * -0.6);
        }
        if (l.getY2() > canvasHeight) {
            l.setY2(canvasHeight);
            l.setYVelocity(l.getYVelocity() * -0.6);
        }
        if (l.getY() < 0) {
            l.setY(-l.getY());
            l.setYVelocity(l.getYVelocity() * -0.6);
        }
        if (l.getY2() < 0) {
            l.setY2(-l.getY2());
            l.setYVelocity(l.getYVelocity() * -0.6);
        }
        if (l.getX2() > canvasWidth) {
            l.setX2(canvasWidth);
            l.setXVelocity(l.getXVelocity() * -0.6);
        }
        if (l.getX() < 0) {
            l.setX(-l.getX());
            l.setXVelocity(l.getXVelocity() * -0.6);
        }
    }

    public void start(int canvasWidth, int canvasHeight) {
        this.canvasWidth = canvasWidth;
        this.canvasHeight = canvasHeight;

        for (Drawable d : drawables) {
            d.setYVelocity(d.getYVelocity() + gravity);
            if (d instanceof Circle) {
                manage((Circle) d);
            } else if (d instanceof Rectangle) {
                manage((Rectangle) d);
            } else if (d instanceof Text) {
                manage((Text) d);
            } else if (d instanceof Triangle && !(d instanceof RightTriangle)) {
                manage((Triangle) d);
            } else if (d instanceof RightTriangle) {
                manage((RightTriangle) d);
            } else if (d instanceof Line) {
                manage((Line) d);
            }
        }
    }
}

File: app/src/main/java/com/util/Gravity.kt
package com.util

import com.drawable.*

@Suppress("unused", "MemberVisibilityCanBePrivate")
class Gravity {
    var drawables: FinalArray<Drawable>? = null
        private set
    var gravity: Double = 0.0
    private var canvasWidth: Int = 0
    private var canvasHeight: Int = 0

    constructor(drawable: Drawable, gravity: Double) : this(arrayOf<Drawable>(drawable), gravity)

    constructor(drawables: Array<Drawable>, gravity: Double) {
        this.drawables = FinalArray.from(drawables)
        this.gravity = gravity
    }

    constructor(drawables: List<Drawable>, gravity: Double) {
        this.drawables = FinalArray.from(drawables.toTypedArray())
        this.gravity = gravity
    }

    private fun manage(c: Circle) {
        if (c.y + c.radius > canvasHeight) {
            c.y = canvasHeight - c.radius
            c.yVelocity = c.yVelocity * -0.6
        }
        if (c.y - c.radius < 0) {
            c.y = c.radius
            c.yVelocity = c.yVelocity * -0.6
        }
        if (c.x + c.radius > canvasWidth) {
            c.x = canvasWidth - c.radius
            c.xVelocity = c.xVelocity * -0.6
        }
        if (c.x - c.radius < 0) {
            c.x = c.radius
            c.xVelocity = c.xVelocity * -0.6
        }
    }

    private fun manage(r: Rectangle) {
        if (r.y + r.height > canvasHeight) {
            r.y = canvasHeight - r.height.toDouble()
            r.yVelocity = r.yVelocity * -0.6
        }
        if (r.y < 0) {
            r.y = -r.y
            r.yVelocity = r.yVelocity * -0.6
        }
        if (r.x + r.width > canvasWidth) {
            r.x = canvasWidth - r.width.toDouble()
            r.xVelocity = r.xVelocity * -0.6
        }
        if (r.x < 0) {
            r.x = -r.x
            r.xVelocity = r.xVelocity * -0.6
        }
    }

    private fun manage(t: Text) {
        if (t.y + t.textSize > canvasHeight + t.textSize * 2) {
            t.y = canvasHeight - t.textSize * 2
            t.yVelocity = t.yVelocity * -0.6
        }
        if (t.y < 0) {
            t.y = t.y + t.textSize / 2
            t.yVelocity = t.yVelocity * -0.6
        }
        if (t.x + t.textSize > canvasWidth + t.textSize / 2 - t.text.length * 16) {
            t.x = canvasWidth.toDouble() - t.textSize / 2 - (t.text.length * 16).toDouble()
            t.xVelocity = t.xVelocity * -0.6
        }
        if (t.x < 0) {
            t.x = t.x + t.textSize / 2
            t.xVelocity = t.xVelocity * -0.6
        }
    }

    private fun manage(t: Triangle) {
        if (t.y > canvasHeight) {
            t.y = canvasHeight.toDouble()
            t.yVelocity = t.yVelocity * -0.6
        }
        if (t.y - t.height < 0) {
            t.y = t.height
            t.yVelocity = t.yVelocity * -0.6
        }
        if (t.x + t.getSideLengths()[2] > canvasWidth) {
            t.x = canvasWidth - t.getSideLengths()[2]
            t.xVelocity = t.xVelocity * -0.6
        }
        if (t.x < 0) {
            t.x = 0.0
            t.xVelocity = t.xVelocity * -0.6
        }
    }

    private fun manage(t: RightTriangle) {
        if (t.y > canvasHeight) {
            t.y = canvasHeight.toDouble()
            t.yVelocity = t.yVelocity * -0.6
        }
        if (t.y - t.height < 0) {
            t.y = t.height
            t.yVelocity = t.yVelocity * -0.6
        }
        if (t.x + t.getSideLengths()[1] > canvasWidth) {
            t.x = canvasWidth - t.getSideLengths()[1]
            t.xVelocity = t.xVelocity * -0.6
        }
        if (t.x < 0) {
            t.x = 0.0
            t.xVelocity = t.xVelocity * -0.6
        }
    }

    private fun manage(l: Line) {
        if (l.y > canvasHeight) {
            l.y = canvasHeight.toDouble()
            l.yVelocity = l.yVelocity * -0.6
        }
        if (l.y2 > canvasHeight) {
            l.y2 = canvasHeight.toDouble()
            l.yVelocity = l.yVelocity * -0.6
        }
        if (l.y < 0) {
            l.y = -l.y
            l.yVelocity = l.yVelocity * -0.6
        }
        if (l.y2 < 0) {
            l.y2 = -l.y2
            l.yVelocity = l.yVelocity * -0.6
        }
        if (l.x2 > canvasWidth) {
            l.x2 = canvasWidth.toDouble()
            l.xVelocity = l.xVelocity * -0.6
        }
        if (l.x < 0) {
            l.x = -l.x
            l.xVelocity = l.xVelocity * -0.6
        }
    }

    fun start(canvasWidth: Int, canvasHeight: Int) {
        this.canvasWidth = canvasWidth
        this.canvasHeight = canvasHeight

        for (d in drawables!!) {
            d.yVelocity += gravity

            //Compiler automatically does a downcast here
            if (d is Circle) {
                manage(d)
            } else if (d is Rectangle) {
                manage(d)
            } else if (d is Text) {
                manage(d)
            } else if (d is Triangle && d !is RightTriangle) {
                manage(d)
            } else if (d is RightTriangle) {
                manage(d)
            } else if (d is Line) {
                manage(d)
            }

        }
    }
}